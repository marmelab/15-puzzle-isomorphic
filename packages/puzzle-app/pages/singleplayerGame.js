import React, { Component } from 'react';

import { defaultPuzzleSize } from '../src/config';

import { initGame, moveTile } from 'puzzle-core';

import Block from '../src/components/layout/Block';
import Page from '../src/components/layout/Page';
import Section from '../src/components/layout/Section';
import Game from '../src/components/puzzle/Game';

import ChangeSizeModal from '../src/components/puzzle/ChangeSizeModal';

export const title = (isLoading, isVictory, turn) => {
    if (isLoading) {
        return 'Building a new game';
    }
    if (isVictory) {
        return `Congratulations, you have solved the puzzle in ${turn} turns!`;
    }
    if (turn === 0) {
        return 'Start the game by moving a tile';
    }
    return `${turn} moves`;
};

export default class SinglePlayerGame extends Component {
    state = {
        currentGrid: [],
        isLoading: true,
        isVictory: false,
        resolvedGrid: [],
        turn: -1,
    };

    handleClickTile = tile => {
        const { currentGrid, resolvedGrid, turn } = this.state;
        try {
            const move = moveTile({ currentGrid, resolvedGrid, turn }, tile);
            this.setState({
                currentGrid: move.currentGrid,
                isVictory: move.isVictory,
                turn: move.turn,
            });
        } catch (error) {
            console.error(error);
            // TODO : catch the findTileByValue and the move errors in order to display them to the user.
        }
    };

    buildGame = async size => {
        this.setState({
            isLoading: true,
        });
        const { currentGrid, resolvedGrid, turn } = await initGame(size);
        this.setState({
            currentGrid,
            isLoading: false,
            isVictory: false,
            resolvedGrid,
            turn,
        });
    };

    handleOnCloseModal = newSize => {
        if (!Number.isInteger(newSize)) {
            return;
        }
        this.buildGame(newSize);
    };

    componentWillMount() {
        this.buildGame(defaultPuzzleSize);
    }

    render() {
        const {
            currentGrid,
            isLoading,
            isVictory,
            resolvedGrid,
            turn,
        } = this.state;

        return (
            <Page navTitle="Singleplayer" title="Singleplayer | 15 puzzle">
                <Section>
                    <Block
                        isLoading={isLoading}
                        title={title(isLoading, isVictory, turn)}
                    >
                        <Game
                            currentGrid={currentGrid}
                            isVictory={isVictory}
                            onClickTile={this.handleClickTile}
                            resolvedGrid={resolvedGrid}
                        />
                    </Block>
                    {turn <= 0 && (
                        <ChangeSizeModal
                            onClose={this.handleOnCloseModal}
                            size={resolvedGrid.length}
                        />
                    )}
                </Section>
            </Page>
        );
    }
}
