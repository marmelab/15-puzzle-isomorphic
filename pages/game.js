import React, { Component } from 'react';
import PropTypes from 'prop-types';

import config from '../src/config';

import Bloc from '../src/components/bloc';
import Button from '../src/components/button';
import { ShowWhenOnline } from '../src/components/detectOffline';
import Grid from '../src/components/grid';
import Page from '../src/components/page';
import Row from '../src/components/row';
import Section from '../src/components/section';

import {
    areGridsEquals,
    buildGrid,
    findTileByValue,
    move,
} from '../src/core/game';
import { shuffle } from '../src/core/shuffler';

import { suggestFactory } from '../src/services/suggestMoveService';

export const title = (isLoading, isVictory, turn) => {
    if (isLoading) {
        return 'Building a new game';
    }
    if (isVictory) {
        return `Congratulations, you have solved the puzzle in ${turn} turns!`;
    }
    return `Turn ${turn}`;
};

export default class Game extends Component {
    state = {
        currentGrid: [],
        isLoading: true,
        isVictory: false,
        loadingAdvice: false,
        resolvedGrid: [],
        suggestedTile: 0,
        turn: -1,
    };

    static getInitialProps = ({ query }) => ({
        size: query.size || config.defaultPuzzleSize,
    });

    static propTypes = {
        size: PropTypes.number.isRequired,
    };

    handleClickTile = tile => {
        const { currentGrid, resolvedGrid, turn } = this.state;

        try {
            const coordsTileToMove = findTileByValue(currentGrid, tile);
            const newCurrentGrid = move(currentGrid, coordsTileToMove);
            const isVictory = areGridsEquals(newCurrentGrid, resolvedGrid);

            const newState = {
                currentGrid: newCurrentGrid,
                isVictory,
                suggestedTile: 0,
                turn: turn + 1,
            };

            this.setState(newState);
        } catch (error) {
            console.error(error);
            // TODO : catch the findTileByValue and the move errors in order to display them to the user.
        }
    };

    requestSuggest = async () => {
        const { currentGrid, resolvedGrid } = this.state;
        this.setState({ loadingAdvice: true });
        try {
            const { Tile } = await suggestFactory()(currentGrid, resolvedGrid);
            this.setState({
                loadingAdvice: false,
                suggestedTile: Tile,
            });
        } catch (error) {
            console.error(error);
            // TODO : catch the errors in order to display them to the user.
        }
    };

    handleClickSuggest = () => {
        const { loadingAdvice } = this.state;
        if (loadingAdvice) {
            return;
        }
        this.requestSuggest();
    };

    buildGame = async () => {
        const { size } = this.props;

        let resolvedGrid = buildGrid(size);
        let currentGrid = await shuffle(resolvedGrid);

        this.setState({
            isLoading: false,
            currentGrid,
            resolvedGrid,
            turn: 0,
            isVictory: false,
        });
    };

    componentWillMount() {
        this.buildGame();
    }

    render() {
        const {
            currentGrid,
            isLoading,
            isVictory,
            loadingAdvice,
            resolvedGrid,
            suggestedTile,
            turn,
        } = this.state;

        return (
            <Page>
                <Section>
                    <Bloc
                        title={title(isLoading, isVictory, turn)}
                        isLoading={isLoading}
                    >
                        {currentGrid && (
                            <Grid
                                onClick={this.handleClickTile}
                                grid={currentGrid}
                                resolvedGrid={resolvedGrid}
                                readOnly={isVictory}
                            />
                        )}
                        {loadingAdvice && <p>Looking for an advice...</p>}
                        {!loadingAdvice &&
                            suggestedTile !== 0 && (
                                <p>You could move the tile {suggestedTile}</p>
                            )}
                    </Bloc>
                </Section>
                <Section>
                    <Row>
                        <div className="buttons-wrapper">
                            <Button
                                icon="keyboard_return"
                                color="red"
                                label="Back to home"
                                route="index"
                            />
                            {!isLoading &&
                                !isVictory && (
                                    <ShowWhenOnline>
                                        <Button
                                            icon="help_outline"
                                            label="Ask for help"
                                            onClick={this.handleClickSuggest}
                                        />
                                    </ShowWhenOnline>
                                )}
                        </div>
                    </Row>
                </Section>
            </Page>
        );
    }
}
