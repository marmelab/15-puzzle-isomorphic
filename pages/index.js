import React, { Component } from 'react';
import CustomHead from '../src/components/head';
import Nav from '../src/components/nav';

export default class Index extends Component {
    handleClickSingleGame = () => {};

    render() {
        return (
            <div>
                <CustomHead />
                <Nav />
                <section className="container">
                    <div className="section">
                        <div className="bloc z-depth-2">
                            <div className="col s12">
                                <h5>Welcome to the 15 puzzle game!</h5>
                            </div>
                            <div className="col s12">
                                <img
                                    className="responsive-img"
                                    src="/static/images/banner.jpg"
                                    alt="15 puzzle picture"
                                />
                            </div>
                        </div>
                    </div>
                    <div className="section">
                        <div className="row">
                            <div className="col s12">
                                <button
                                    onClick={this.handleClickSingleGame}
                                    className="btn blue"
                                >
                                    <i className="material-icons left">
                                        play_circle_outline
                                    </i>Single game
                                </button>
                            </div>
                        </div>
                    </div>
                </section>
            </div>
        );
    }
}
