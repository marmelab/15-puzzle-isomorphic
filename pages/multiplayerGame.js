import React, { Component } from 'react';
import PropTypes from 'prop-types';

import config from '../src/config';
import { Router } from '../src/routes';

import Bloc from '../src/components/bloc';
import Button from '../src/components/button';
import Grid from '../src/components/grid';
import Page from '../src/components/page';
import Row from '../src/components/row';
import Section from '../src/components/section';

import {
    gameFactory,
    joinFactory,
    moveFactory,
} from '../src/services/multiplayerGameService';

export const title = (
    id,
    isLoading,
    isVictory,
    isWaitingPlayer,
    isWinner,
    turn,
) => {
    if (isLoading && !isWaitingPlayer) {
        return `Loading the game #${id}`;
    }
    if (isWaitingPlayer) {
        return `Waiting for another player in game #${id}`;
    }
    if (isWinner) {
        if (isVictory) {
            return `Congratulations, you have solved the puzzle in ${
                turn
            } turns!`;
        }
        return `Sorry, you opponent solved the puzzle in ${turn} turns!`;
    }
    return `Turn ${turn}`;
};

export default class MultiplayerGame extends Component {
    state = {
        isLoading: true,
        isWaitingPlayer: false,
        id: -1,
        otherPlayerId: -1,
        isMultiplayer: false,
        token: '',
        currentGrid: [],
        turn: -1,
        winnerId: -1,
    };

    static propTypes = {
        id: PropTypes.string.isRequired,
        token: PropTypes.string,
    };

    static getInitialProps = async ({ query }) => {
        return {
            id: query.id,
            token: query.token || null,
        };
    };

    waitForOtherPlayer = async (id, token) => {
        const { otherPlayer } = await gameFactory()(id, token);
        if (otherPlayer) {
            return Promise.resolve(otherPlayer);
        }

        await new Promise(resolve => {
            setTimeout(resolve, config.refreshDuration);
        });

        return this.waitForOtherPlayer(id, token);
    };

    requestGame = async (id, token) => {
        try {
            let {
                isMultiplayer,
                currentPlayer,
                otherPlayer,
                winner,
            } = await gameFactory()(id, token);

            if (isMultiplayer && !otherPlayer) {
                this.setState({ id, isWaitingPlayer: true });
                otherPlayer = await this.waitForOtherPlayer(id, token);
                this.setState({ isWaitingPlayer: false });
            }

            let newState = {
                isLoading: false,
                id,
                token,
                playerId: currentPlayer.id,
                isMultiplayer,
                currentGrid: currentPlayer.currentGrid,
                turn: currentPlayer.turn,
                winnerId: winner !== null ? winner.id : -1,
            };

            if (isMultiplayer && otherPlayer) {
                newState.otherPlayerId = otherPlayer.id;
            }
            this.setState(newState);
        } catch (error) {
            console.error(error);
            // TODO: catch the server error in order to display it to the user.
            Router.pushRoute('multiplayer_games');
        }
    };

    requestMove = async tile => {
        try {
            const { id, token } = this.state;
            const { currentPlayer, winner } = await moveFactory()(
                id,
                token,
                tile,
            );
            this.setState({
                currentGrid: currentPlayer.currentGrid,
                turn: currentPlayer.turn,
                winnerId: winner !== null ? winner.id : -1,
            });
        } catch (error) {
            console.error(error);
            // TODO: catch the server error in order to display it to the user.
        }
    };

    requestJoin = async id => {
        try {
            const { token } = await joinFactory()(id);
            Router.pushRoute('multiplayer_game', { id, token });
        } catch (error) {
            console.error(error);
            // TODO: catch the server error in order to display it to the user.
        }
    };

    handleClick = tile => {
        this.requestMove(tile);
    };

    componentWillMount() {
        const { id, token } = this.props;

        if (id === -1) {
            Router.pushRoute('multiplayer_games');
            return;
        }
        if (!token) {
            this.requestJoin(id);
        }
        this.requestGame(id, token);
    }

    render() {
        const {
            id,
            currentGrid,
            isLoading,
            isWaitingPlayer,
            playerId,
            turn,
            winnerId,
        } = this.state;
        const isWinner = winnerId !== -1;
        const isVictory = isWinner && winnerId === playerId;

        return (
            <Page>
                <Section>
                    <Bloc
                        title={title(
                            id,
                            isLoading,
                            isVictory,
                            isWaitingPlayer,
                            isWinner,
                            turn,
                        )}
                        isLoading={isLoading}
                    >
                        <Grid
                            onClick={this.handleClick}
                            grid={currentGrid}
                            readOnly={isVictory}
                        />
                    </Bloc>
                </Section>
                <Section>
                    <Row>
                        <div className="buttons-wrapper">
                            <Button
                                icon="keyboard_return"
                                color="red"
                                label="Back to home"
                                route="multiplayer_games"
                            />
                        </div>
                    </Row>
                </Section>
            </Page>
        );
    }
}
