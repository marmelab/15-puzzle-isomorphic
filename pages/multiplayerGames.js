import React, { Component } from 'react';

import Router from '../src/routes';

import Bloc from '../src/components/bloc';
import ListGames from '../src/components/listGames';

import Page from '../src/components/page';
import Section from '../src/components/section';

import { games } from '../src/services/multiplayerGameService';

export default class MultiplayerGames extends Component {
    state = {
        isLoading: true,
        openMultiplayerGames: [1, 2],
    };

    requestGames = async () => {
        try {
            const { gameIds } = await games()();
            this.setState({
                isLoading: false,
                openMultiplayerGames: Object.values(gameIds).map(
                    game => game.id,
                ),
            });
        } catch (error) {
            console.error(error);
            // TODO: catch the server error in order to display it to the user.
            Router.pushRoute('index');
        }
    };

    requestJoin = async () => {};

    handleOnGameSelected = async id => {
        this.requestJoin(id);
    };

    componentWillMount() {
        this.requestGames();
    }

    render() {
        const { isLoading, openMultiplayerGames } = this.state;

        return (
            <Page>
                <Section>
                    <Bloc title="Open multiplayer games" isLoading={isLoading}>
                        <ListGames
                            games={openMultiplayerGames}
                            onGameSelected={this.handleOnGameSelected}
                        />
                    </Bloc>
                </Section>
            </Page>
        );
    }
}
