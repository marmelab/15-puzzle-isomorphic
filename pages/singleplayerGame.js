import React, { Component } from 'react';

import { defaultPuzzleSize } from '../src/config';

import { initGame, moveTile } from '../src/core/game';

import Block from '../src/components/layout/Block';
import Page from '../src/components/layout/Page';
import Section from '../src/components/layout/Section';
import Game from '../src/components/puzzle/Game';

import ChangeSizeModal from '../src/components/puzzle/ChangeSizeModal';

export const title = (isLoading, isVictory, turn) => {
    if (isLoading) {
        return 'Building a new game';
    }
    if (isVictory) {
        return `Congratulations, you have solved the puzzle in ${turn} turns!`;
    }
    if (turn === 0) {
        return 'Start the game by moving a tile';
    }
    return `${turn} moves`;
};

export default class SinglePlayerGame extends Component {
    state = {
        currentGrid: [],
        isLoading: true,
        isVictory: false,
        resolvedGrid: [],
        size: 4,
        turn: -1,
    };

    handleClickTile = tile => {
        const { currentGrid, resolvedGrid, turn } = this.state;

        try {
            const move = moveTile({ currentGrid, resolvedGrid, turn }, tile);

            this.setState({
                currentGrid: move.currentGrid,
                isVictory: move.isVictory,
                suggestedTile: 0,
                turn: move.turn,
            });
        } catch (error) {
            console.error(error);
            // TODO : catch the findTileByValue and the move errors in order to display them to the user.
        }
    };

    buildGame = async () => {
        this.setState({
            isLoading: true,
        });
        const { currentGrid, resolvedGrid, turn } = await initGame(
            this.state.size,
        );
        this.setState({
            currentGrid,
            isLoading: false,
            isVictory: false,
            resolvedGrid,
            turn,
        });
    };

    handleOnCloseModal = newSize => {
        if (!Number.isInteger(newSize)) {
            return;
        }
        this.setState({ size: newSize });
        this.buildGame();
    };

    componentWillMount() {
        this.buildGame();
    }

    render() {
        const {
            currentGrid,
            isLoading,
            isVictory,
            resolvedGrid,
            size,
            suggestedTile,
            turn,
        } = this.state;

        return (
            <Page navTitle="Singleplayer" title="Singleplayer | 15 puzzle">
                <Section>
                    <Block
                        isLoading={isLoading}
                        title={title(isLoading, isVictory, turn)}
                    >
                        <Game
                            currentGrid={currentGrid}
                            isVictory={isVictory}
                            onClickTile={this.handleClickTile}
                            resolvedGrid={resolvedGrid}
                            suggestedTile={suggestedTile}
                        />
                    </Block>
                    {turn <= 0 && (
                        <ChangeSizeModal
                            onClose={this.handleOnCloseModal}
                            size={size}
                        />
                    )}
                </Section>
            </Page>
        );
    }
}
