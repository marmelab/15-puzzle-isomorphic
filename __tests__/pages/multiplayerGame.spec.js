import { title } from '../../pages/multiplayerGame';

describe('title', () => {
    test('should render a turn', () => {
        const expectedTitle = 'Turn 12';
        expect(title(1, false, false, false, false, 12)).toEqual(expectedTitle);
    });

    test('should render a victory', () => {
        const expectedTitle =
            'Congratulations, you have solved the puzzle in 12 turns!';
        expect(title(1, false, true, false, true, 12)).toEqual(expectedTitle);
    });

    test('should render a victory from the opponent', () => {
        const expectedTitle =
            'Sorry, you opponent solved the puzzle in 12 turns!';
        expect(title(1, false, false, false, true, 12)).toEqual(expectedTitle);
    });

    test('should render a waiting for a player message', () => {
        const expectedTitle = 'Waiting for another player in game #1';
        expect(title(1, false, false, true, false, 0)).toEqual(expectedTitle);
    });

    test('should render a loading', () => {
        const expectedTitle = 'Loading the game #1';
        expect(title(1, true, false, false, false, 0)).toEqual(expectedTitle);
    });
});
