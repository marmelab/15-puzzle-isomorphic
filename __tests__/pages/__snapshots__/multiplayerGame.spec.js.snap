// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiplayerGames should render the MultiplayerGame correctly 1`] = `
<div>
  <nav
    className="blue lighten-1"
    role="navigation"
  >
    <div
      className="nav-wrapper container"
    >
      <a
        className="brand-logo"
        href="/"
        onClick={[Function]}
      >
        15 Puzzle Isomorphic!
      </a>
    </div>
  </nav>
  <section
    className="container"
  >
    <p>
      loading...
    </p>
    <p>
      loading...
    </p>
  </section>
</div>
`;
